tests/kernel/timer/timer_behavior/src/tick_timer_train.c

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/time_units.h>
+#include <zephyr/tc_util.h>
+#include <zephyr/ztest.h>
+
+#define TIMERS 1
+#define TEST_SECONDS 10
+#define MAX_CALLBACKS (CONFIG_SYS_CLOCK_TICKS_PER_SEC*TEST_SECONDS)
+
+struct timer_wrapper {
+	uint32_t callbacks;
+	struct k_timer tm;
+	uint32_t last_isr;
+	int64_t last_scheduled;
+	uint32_t max_delta;
+};
+
+K_SEM_DEFINE(timers_sem, 0, K_SEM_MAX_LIMIT);
+
+struct timer_wrapper timers[TIMERS];
+
+void tm_fn(struct k_timer *tm)
+{
+	struct timer_wrapper *tm_wrap =
+		CONTAINER_OF(tm, struct timer_wrapper, tm);
+	uint32_t now = k_cycle_get_32();
+
+	if (tm_wrap->last_isr != 0) {
+		uint32_t delta = now - tm_wrap->last_isr;
+
+		tm_wrap->max_delta = delta > tm_wrap->max_delta ? delta : tm_wrap->max_delta;
+	}
+	tm_wrap->last_isr = now;
+	tm_wrap->callbacks++;
+	if (tm_wrap->callbacks >= MAX_CALLBACKS) {
+		k_timer_stop(tm);
+		k_sem_give(&timers_sem);
+	} else {
+		int64_t next = tm_wrap->last_scheduled + TIMERS;
+
+		tm_wrap->last_scheduled = next;
+		k_timer_start(tm, K_TIMEOUT_ABS_TICKS(next), K_NO_WAIT);
+	}
+}
+
+
+/**
+ * @brief Test timers can be scheduled 1 tick apart without issues
+ *
+ * Schedules timers with absolute scheduling with a 1 tick
+ * period. Measures the total time elapsed and tries to run
+ * some fake busy work while doing so. If the print outs don't show up or
+ * the timer train is late to the station, the test fails.
+ */
+ZTEST(timer_behavior, test_one_tick_timer_train)
+{
+	const uint32_t max_time = TEST_SECONDS*1000 + 1000;
+
+	TC_PRINT("Initializing %u Timers, Tick Rate %uHz, Expecting %u callbacks in %u ms\n",
+		 TIMERS, CONFIG_SYS_CLOCK_TICKS_PER_SEC, MAX_CALLBACKS, max_time);
+
+	for (int i = 0; i < TIMERS; i++) {
+		k_timer_init(&timers[i].tm, tm_fn, NULL);
+		timers[i].max_delta = 0;
+	}
+
+	TC_PRINT("Starting Timers with Skews\n");
+	int64_t tick = k_uptime_ticks();
+
+	for (int i = 0; i < TIMERS; i++) {
+		timers[i].last_scheduled = tick+1;
+		k_timer_start(&timers[i].tm, K_TIMEOUT_ABS_TICKS(tick + 1), K_NO_WAIT);
+	}
+
+#ifdef CONFIG_TIMER_HAS_64BIT_CYCLE_COUNTER
+	uint64_t start_cycle = k_cycle_get_64();
+#else
+	uint32_t start_time_ms = k_uptime_get();
+#endif
+
+	uint32_t remaining_timers = TIMERS;
+
+	/* Do work in the meantime, proving there's enough time to do other things */
+	uint32_t busy_loops = 0;
+
+	while (true) {
+		while (k_sem_take(&timers_sem, K_NO_WAIT) == 0) {
+			remaining_timers--;
+
+		}
+		if (remaining_timers == 0) {
+			break;
+		}
```

Comment:
This loop took me a while to figure out.  How about this, which is simpler and easier to understand:

```c
while (remaining_timers > 0) {
    if (k_sem_take(&timers_sem, K_NO_WAIT) == 0) {
        remaining_timers--;
    }
    busy_loops++;
    k_busy_wait(250000);
}

Reply:
Sorry, didn't update the comment for the next one I wrote: you'd want the busy_wait() in this version to be much faster (or just absent, it should be OK to spin on a semaphore).

Reply:
Or another option would be to spawn one low priority thread that just does a "`while(true);`" and just do this with for loop over the known number of timer events.

Reply:
I want to validate that the thread is in fact running and that it isn't really prevented from running for any significant duration, the time period isn't a big deal. See how busy_loops is used below to validate that yes, the thread ran, and it ran for a reasonable amount of time despite the timers being maxed out.

Reply:
The way the test is setup, I'm testing that timers and its machinery don't take more than give or take 10% of the time available even if scheduled on every tick. Perhaps that's too generous to threads, but that's probably reasonable given most work is really meant to happen in threads and spending more time in timer machinery than threads is a net negative in the worst case scenario setup here. Hopefully that makes it clearer, I'm not sure the thread spawned up with while(true) really shows that where as the k_busy_wait and count do.